src/Processors/Chunk.cpp

```diff
@@ -19,8 +19,8 @@ Chunk::Chunk(DB::Columns columns_, UInt64 num_rows_) : columns(std::move(columns
     checkNumRowsIsConsistent();
 }
 
-Chunk::Chunk(Columns columns_, UInt64 num_rows_, ChunkInfoPtr chunk_info_)
-    : columns(std::move(columns_)), num_rows(num_rows_), chunk_info(std::move(chunk_info_))
+Chunk::Chunk(Columns columns_, UInt64 num_rows_, ChunkInfoPtr chunk_info_, int part_level_)
```

Comment:
`chunk_info` feels natural for this.

Reply:
Initially I wanted to use `ChunkInfo`, but then I found that later `hasEqualSortColumnsWith` is called very frequent and it can accumulate the cost of `dynamic_cast`. I think std::optional as in your is a better choice.

Reply:
Where is the `dynamic_cast`?

Reply:
I mean if we create a new type of `ChunkInfo` to carry `part_level`, then when calling `mayContainRowsWithSamePrimaryKeys` we need to dynamic cast `chunk_info` to our expected type.

Reply:
Sorry not `dynamic_cast` but `typeid_cast` to be more accurate. But even with `typeid_cast` we need to check `typeid` before casting.
And we have number of calls to `mayContainRowsWithSamePrimaryKeys` == number of calls to `hasEqualSortColumnsWith` calls == number of read rows. That's quite a lot.

Reply:
We can introduce a new type of ChunkInfo, but instead of doing dynamic_cast, add a virtual method `getPartLevel() const`